### ISSUE TITLE
Under what conditions does the MLLM planner take effect?

### ISSUE THREAD

**Issue by tyfeld** [bytedance/Bernini#28]
Thanks for your inspiring work! We're trying to understand when the MLLM planner in `ByteDance/Bernini-Diffusers` is supposed to drive the output. Across every task we tried (i2i / t2v / v2v with the official cases) we cannot make the planner produce any measurable change — the renderer behaves as if T5 were the only conditioning signal. We'd like to know:
- on which task types / case patterns / inference flags is the planner expected to be the dominant contributor?
- is there a known-good case where switching the planner off should visibly degrade the output?

Below is what we measured, in case it helps narrow down whether this is a "you're testing the wrong cases" issue or a "the released checkpoint isn't using the planner" issue.

## Evidence 1 — cross-attention V projection is ~0 for the planner segment

Instrumented `WanAttnProcessor2_0.__call__` to dump K/V norms on the first denoising step, splitting the 512-token KV sequence into `[T5 real | planner real | pad zeros]` (i2i baseline gives `86 / 222 / 204`). Mean over heads:

| layer | V_T5 | **V_planner** | **V_pad** |
|---|---|---|---|
|  0 | 3.84 | **0.016** | **0.015** |
| 20 | 2.74 | **0.014** | **0.012** |
| 39 | 2.05 | **0.008** | **0.007** |

K norms are similar across all three segments (~10), but **V_planner ≈ V_pad ≈ `‖b_v‖`** and ~250× smaller than V_T5. It looks like `W_v · connector.for_gen(planner) ≈ b_v` no matter what the planner produced. Attention mass distributes ~40% to planner / ~50% to pad / ~10% to T5, but multiplied by near-zero V the cross-attn output is dominated by T5.

## Evidence 2 — behavioural ablations (same seed=42)

I added env-var hooks before the `torch.cat([t5_embeds, cond_embeds_*])` step to zero one side or add noise:

| Ablation | i2i delete-bicycle | t2v polar bear | v2v add-snowman |
|---|---|---|---|
| baseline | clean edit | correct scene | correct edit |
| `NO_PLANNER=1` (zero planner cond_embeds) | identical | identical | **PSNR 36.7 dB vs baseline** |
| `NO_T5=1` (zero T5) | breaks | breaks | (skipped) |
| `PLANNER_NOISE_SCALE=K`, K ∈ {0.1, 1, 10, 30} | identical at every K | — | — |
| MLLM gets "add red car, beach, palm trees" while T5 gets "delete bicycle" | bicycle deleted, no car / beach | — | — |

<img width="1258" height="583" alt="Image" src="https://github.com/user-attachments/assets/36b68652-3e1d-4072-b732-a7a7fed0bc15" />
<img width="1524" height="366" alt="Image" src="https://github.com/user-attachments/assets/d9b26283-133d-485d-a166-ef1c13bdb980" />

We also tried injecting noise into the planner output and feeding the MLLM a deliberately wrong prompt — neither changed the result.

<img width="1258" height="638" alt="Image" src="https://github.com/user-attachments/assets/5b50836f-b6dd-451b-8562-12a1e729bbb5" />

## Questions

1. **What case should we test where the planner is expected to dominate?** A known-good diagnostic prompt / video / reference image that you have observed planner-vs-no-planner deltas on would let us verify our setup.
2. Is there a showcase about image reasoning/reflection such as that in the paper?

**Comment by cybe42** [comment]
https://github.com/Comfy-Org/ComfyUI/issues/14416#issuecomment-4753228739

**Comment by mzsun-triz** [comment]
Thank you for the detailed investigation. Your observations provide a useful perspective on the conditioning behavior of **Bernini V1**. We appreciate the careful instrumentation and ablation experiments, which help illustrate how the different conditioning signals contribute to the model output.

## Analysis of the conditioning behavior

The available evidence suggests that the relative influence of MLLM conditions is weaker than that of the T5 condition in some layers and evaluation cases. This behavior may be related to the original training dynamics: the connector was initialized near zero and jointly optimized with the full model from the beginning. Under this setup, the renderer may learn to rely more heavily on the already well-aligned T5 features during the early stage of training, while adapting more slowly to the newly introduced MLLM conditions.

The effect is particularly visible in the reported post-RMSNorm value projections. Although the MLLM-text and target-ViT tokens can receive non-trivial attention probabilities, their value magnitudes in the examined layers are substantially smaller than those of the T5 tokens. This may reduce their effective contribution to the cross-attention output and make planner-on versus planner-off differences less apparent in some cases. Bernini V2 uses a revised training strategy designed to strengthen the influence of the MLLM conditions.

## Updated training recipe in Bernini V2

Bernini V2 adopts a revised training recipe that warms up the connector before co-training:

1. **Connector warm-up:** We first freeze the DiT renderer and train the connector to align the MLLM outputs with the renderer's conditioning space. We warm up the connector on T2I/T2V data for 2K steps, and then include I2I/V2V data for another 4K steps.
2. **Full co-training:** We initialize from the warmed-up connector, unfreeze all parameters, and jointly train the MLLM–connector–DiT system.

The main difference is that the connector learns a meaningful projection before the renderer is jointly optimized. This gives the renderer an opportunity to learn from the MLLM conditions from an earlier stage of full co-training, rather than relying primarily on the already well-aligned T5 features.

## V-norm comparison

Below are representative post-RMSNorm V norms measured from the first cross-attention block using the same input, denoising step, and inference configuration:

| Model | T5 V norm | MLLM-text V norm | target-ViT V norm |
| --- | ---: | ---: | ---: |
| Bernini V1 | 4.145 | 0.016 | 0.016 |
| Bernini V2 | 4.129 | 5.273 | 8.011 |

In Bernini V1, the MLLM-text and target-ViT V norms in this examined block are both approximately `0.016`, substantially smaller than the T5 V norm. In Bernini V2, their V norms become comparable to—or larger than—the T5 V norm, giving the corresponding conditions a greater capacity to influence the cross-attention output.

## Quantitative comparison between Bernini V1 and V2

We evaluated Bernini V2 on Bernini-RV2V, OpenS2V, and VBench:

| Model | EditVerse | OpenVE | OpenS2V | VBench | Bernini-V2V (OS) | Bernini-RV2V (OS) |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Bernini V1 | 8.02 | 4.03 | 62.30 | 84.37 | 3.49 | 3.48 |
| Bernini V2 | 8.02 | 3.96 | 63.83 | 84.46 | 3.49 | 3.55 |

Bernini V2 improves Bernini-RV2V from `3.48` to `3.55` and OpenS2V from `62.30` to `63.83`. Its VBench score changes slightly from `84.37` to `84.46`. These results suggest that the revised training recipe strengthens editing-oriented performance while largely preserving general video-generation quality.

## Quantitative conditioning ablations on Bernini V2

To quantify the contributions of the different conditions under the revised training strategy, we performed controlled ablations on the Bernini-V2V benchmark. The renderer initialization, training data, training steps, resolution, and inference protocol were held fixed across all variants:

| ID | Setting | OS ↑ | IF ↑ | VC ↑ | GQ ↑ |
| --- | --- | ---: | ---: | ---: | ---: |
| A | T5 text + source VAE features | 3.15 | 3.36 | 3.28 | 3.45 |
| B | A + MLLM hidden states | 3.21 | 3.44 | 3.11 | 3.47 |
| C | B + target-ViT planning | 3.41 | 3.47 | 3.56 | 3.57 |
| D | C + prompt enhancement | 3.49 | 3.60 | 3.65 | 3.52 |

Comparing A and B, MLLM conditioning improves OS from `3.15` to `3.21`, IF from `3.36` to `3.44`, and GQ from `3.45` to `3.47`, while VC decreases from `3.28` to `3.11`. The mixed result suggests that the MLLM hidden states mainly benefit instruction-level conditioning but do not improve every dimension on their own. Adding target-ViT planning in C raises all four metrics, with the largest gain in VC (`3.11` to `3.56`), indicating that the target-side visual plan complements the MLLM hidden states with stronger visual guidance. Prompt enhancement in D further improves OS, IF, and VC, while GQ changes from `3.57` to `3.52`.

## Quantitative analysis of target-ViT planning alignment

We also measured the alignment between the predicted plan and the ground-truth target-ViT representation on 1,000 samples:

| Training step | MSE ↓ | Cosine similarity ↑ | Pearson correlation ↑ |
| --- | ---: | ---: | ---: |
| 5K | 0.4857 | 0.6571 | 0.6342 |
| 10K | 0.4062 | 0.7191 | 0.7098 |
| 12K | 0.3868 | 0.7331 | 0.7260 |
| 15K | **0.3676** | **0.7476** | **0.7417** |

The consistent improvement across these metrics indicates that the predicted target-ViT representation becomes progressively better aligned with its supervised target during training.

We hope these results clarify the behavior observed in Bernini V1 and the motivation for the revised training strategy. Bernini V2, trained with the revised recipe described above, is now available in the repository, and we recommend it for future evaluation and experimentation. Thank you again for the careful investigation and detailed feedback.
